- document deployment approaches

SSE4 opened the pull request:
closes: #1423 
rewrite of https://github.com/conan-io/conan-io.github.io/pull/88

/cc @lasote @uilianries @danimtb @jgsogo @czoido @memsharded 

pls tell me if the overall structure is good

lasote commented on howtos/deployment.rst:
```diff
@@ -0,0 +1,348 @@
+system package manager
+----------------------
+
+Usually the recommended option is to use system package manager, like making **.rpm** or ***.deb** package(s). Unfortunately, 
```
Recommended by who? I would omit that sentence and the "unfortunately etc"
If all this is going to be on "Integrations" it should look like we are explaining how to create (or at least link to some tool or manual). But off course, we should comment the limitations.

SSE4 replied:
I've made this more neutral, removing all things like `recommended` and `unfortunately`

lasote commented on howtos/deployment.rst:
```diff
@@ -0,0 +1,348 @@
+Usually the recommended option is to use system package manager, like making **.rpm** or ***.deb** package(s). Unfortunately, 
+it's not always an option, as there are few limitations:
+
+- scalability problem - need to create one package per supported distro (likely one for `Ubuntu <https://ubuntu.com/>`_, one for `Arch Linux <https://www.archlinux.org/>`_, etc.), and formats or guidelines for each distro might differ significantly
```
```suggestion
- scalability problem - If you want to target different distros, then you need to create one package per supported distro (likely one for `Ubuntu <https://ubuntu.com/>`_, one for `Arch Linux <https://www.archlinux.org/>`_, etc.), and formats or guidelines for each distro might differ significantly
```

lasote commented on howtos/deployment.rst:
```diff
@@ -0,0 +1,348 @@
+it's not always an option, as there are few limitations:
+
+- scalability problem - need to create one package per supported distro (likely one for `Ubuntu <https://ubuntu.com/>`_, one for `Arch Linux <https://www.archlinux.org/>`_, etc.), and formats or guidelines for each distro might differ significantly
+- it might be needed to (re)compile application N times for each of supported distro, or at least use the lowest version (see concerns about ``glibc`` below)
```
Put this point the first one. Link to https://docs.conan.io/en/latest/extending/custom_settings.html#customizing-settings and explain that it would be needed to customize the settings with, for exampe, the distribution you are targeting to avoid issues with different glibc.

SSE4 replied:
moved and added a link

lasote commented on howtos/deployment.rst:
```diff
@@ -0,0 +1,348 @@
+
+- scalability problem - need to create one package per supported distro (likely one for `Ubuntu <https://ubuntu.com/>`_, one for `Arch Linux <https://www.archlinux.org/>`_, etc.), and formats or guidelines for each distro might differ significantly
+- it might be needed to (re)compile application N times for each of supported distro, or at least use the lowest version (see concerns about ``glibc`` below)
+- sometimes older distributions may lack packages for actual software releases (e.g. boost 1.71 might be missing in `CentOS <https://www.centos.org/?>`_ 6 or 7)
```
I think this point is not so relevant. We are not selling using other systems, if a company has as a requirement to deploy using RPM they know about this.

SSE4 replied:
removed

lasote commented on howtos/deployment.rst:
```diff
@@ -0,0 +1,348 @@
+- scalability problem - need to create one package per supported distro (likely one for `Ubuntu <https://ubuntu.com/>`_, one for `Arch Linux <https://www.archlinux.org/>`_, etc.), and formats or guidelines for each distro might differ significantly
+- it might be needed to (re)compile application N times for each of supported distro, or at least use the lowest version (see concerns about ``glibc`` below)
+- sometimes older distributions may lack packages for actual software releases (e.g. boost 1.71 might be missing in `CentOS <https://www.centos.org/?>`_ 6 or 7)
+- not all systems use package managers, for instance, some embedded systems or `LFS <http://www.linuxfromscratch.org/>`_ distributions
```
Same. Remove

SSE4 replied:
removed

lasote commented on howtos/deployment.rst:
```diff
@@ -0,0 +1,348 @@
+- sometimes older distributions may lack packages for actual software releases (e.g. boost 1.71 might be missing in `CentOS <https://www.centos.org/?>`_ 6 or 7)
+- not all systems use package managers, for instance, some embedded systems or `LFS <http://www.linuxfromscratch.org/>`_ distributions
+
+As a result, there are other ways to help with application deployment.
```
Remove. Do not conclude you shouldn't do that, just present options, objective. This is docs, not a blogpost.

lasote commented on howtos/deployment.rst:
```diff
@@ -0,0 +1,348 @@
+archives (tarballs)
+-------------------
+
+The most straightforward thing to do is to make an archive with application and all of its dependencies and resources, such as zip file, or tarball (``.tar.gz``/``.tgz``, ``.tar.bz``/``.tbz``, ``.tar.xz``/``.txz`` or just ``.tar`` file).
```
```suggestion
You can package your shared libraries, assets, executables etc to a (``.tar.gz``/``.tgz``, ``.tar.bz``/``.tbz``, ``.tar.xz``/``.txz`` or just ``.tar`` file).
```
But I don't think we are providing any useful information here, everyone knows that something can be zipped. maybe remove enterely.

SSE4 replied:
removed

lasote commented on howtos/deployment.rst:
```diff
@@ -0,0 +1,348 @@
+- application is being run
+- finally, the temporary directory removed
+
+Therefore, it transparently appears just like a normal application execution.
```
At least link to a getting started guide or something if we cannot provide an example.

SSE4 replied:
added link to http://xmodulo.com/how-to-create-a-self-extracting-archive-or-installer-in-linux.html

lasote commented on howtos/deployment.rst:
```diff
@@ -0,0 +1,348 @@
+
+Alternatively, ``flatpak`` allows distributing the `single-file <http://docs.flatpak.org/en/latest/single-file-bundles.html>`_ package. Such package, however, cannot be run or installed on its own, it's needed to be imported to the local repository on another machine.
+
+Other
```
```suggestion
Others
```

lasote commented on howtos/deployment.rst:
```diff
@@ -0,0 +1,348 @@
+There are enterprise solutions for deployment, which are recommended to be used for production environments, such as 
+`ansible <https://www.ansible.com/>`_, `chef <https://www.chef.io/application-deployment/>`_ and `puppet <https://puppet.com/>`_.
+
+Challenges
```
Move this section to "https://docs.conan.io/en/latest/devtools/running_packages.html#running-and-deploying-packages" after the deployment one, you can call it "Deployment challenges"

lasote replied:
Link somewhere to the customizing settings too with a sentence that transmit the idea, you can define what is your compatibility mode, customizing settings and generating packages for that specific config, they will be different from other packages with a different glibc for instance.

lasote replied:
And also link back from the customizing settings section to here.

SSE4 replied:
I've split pages and added links back and forward to the customizing settings

SSE4 commented:
- changed the structure, split into https://docs.conan.io/en/latest/devtools/running_packages.html#deployable-packages and https://docs.conan.io/en/latest/integrations.html (new section: `deployment`)
- removed the most of introduction
- comments addressed

lasote commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,227 @@ with:
+Using deploy generator
+----------------------
+
+With the help of so-called :ref:`deploy generator <deploy_generator>`, it's possible to have all dependencies of the application to be copied into the 
```
```suggestion
With the help of the :ref:`deploy generator <deploy_generator>`, it's possible to have all dependencies of the application to be copied into 
```
For the docs, better a plain/boring/simple language.

lasote commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,227 @@ with:
+
+With the help of so-called :ref:`deploy generator <deploy_generator>`, it's possible to have all dependencies of the application to be copied into the 
+a single place for the later repackaging into the desired distribution format.
+For instance, if the application depends on boost, we may not know that it also may require many other 3rt-party libraries, 
```
```suggestion
For instance, if the application depends on boost, we may not know that it also requires many other 3rt-party libraries, 
```

lasote commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,227 @@ with:
+`bzip2 <https://sourceware.org/bzip2/>`_, 
+`lzma <https://tukaani.org/xz/>`_, 
+`zstd <https://facebook.github.io/zstd/>`_, 
+`iconv <https://www.gnu.org/software/libiconv/>`_, etc. It would be nice to collect them all, regardless of options applied to the **boost**.
```
```suggestion
`iconv <https://www.gnu.org/software/libiconv/>`_, etc. 
```

lasote commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,227 @@ with:
+Using json generator
+--------------------
+
+The more advanced approach is to use the :ref:`json generator <json_generator>`. First off, invoke the conan to generate the ``conanbuildinfo.json`` 
```
```suggestion
A more advanced approach is to use the :ref:`json generator <json_generator>`:
```

lasote commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,227 @@ with:
+
+    $ conan install . -g json
+
+The file produced is fully machine-readable and could be used by scripts to prepare the distribution.
```
```suggestion
The `conanbuildinfo.json` file produced is fully machine-readable and could be used by scripts to prepare the distribution.
```

lasote commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,227 @@ with:
+        copy_tree(src_bin_dir, os.path.join(destination, dst_bin_dir))
+
+The advantage over the ``deploy`` generator is fine-grained control: here we copy only binaries and libraries.
+It's also could be easily modified to apply some sort of filtering (e.g. to copy only shared libraries, 
```
I think this is too much. Remove this second example. If the contents are too big people won't read it.

SSE4 replied:
sure

SSE4 replied:
removed

lasote commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,227 @@ with:
+Deployment challenges
+---------------------
+
+C standard library
```
Have you tried to build the pages? you are putting everything at the same section level, right?

SSE4 replied:
changed the formatting to use `~~~~~~~~~~` as 3rd level heading

lasote commented on integrations/deployment.rst:
```diff
@@ -0,0 +1,102 @@
+How to deploy C++ applications with conan
+=========================================
+
+For the time being, it's frequently asked how to deploy C++ application which is built with help of `conan <https://conan.io>`_, 
```
```suggestion
For the time being, it's frequently asked how to deploy C++ application which is built with help of `conan <https://conan.io>`_, 
```
Irrelevant introduction

SSE4 replied:
removed

lasote commented on integrations/deployment.rst:
```diff
@@ -0,0 +1,102 @@
+=========================================
+
+For the time being, it's frequently asked how to deploy C++ application which is built with help of `conan <https://conan.io>`_, 
+and how to generate a deployable artifact after the build.
```
```suggestion

```

lasote commented on integrations/deployment.rst:
```diff
@@ -0,0 +1,102 @@
+
+For the time being, it's frequently asked how to deploy C++ application which is built with help of `conan <https://conan.io>`_, 
+and how to generate a deployable artifact after the build.
+Consider we have a C++ application, with a complex dependency graph, and use conan to manage these dependencies from time to time, 
```
```suggestion

```

lasote commented on integrations/deployment.rst:
```diff
@@ -0,0 +1,102 @@
+For the time being, it's frequently asked how to deploy C++ application which is built with help of `conan <https://conan.io>`_, 
+and how to generate a deployable artifact after the build.
+Consider we have a C++ application, with a complex dependency graph, and use conan to manage these dependencies from time to time, 
+it may be possible to significantly reduce the complexity of deployment via the help of some built-in conan facilities. 
```
```suggestion

```

lasote commented on integrations/deployment.rst:
```diff
@@ -0,0 +1,102 @@
+and how to generate a deployable artifact after the build.
+Consider we have a C++ application, with a complex dependency graph, and use conan to manage these dependencies from time to time, 
+it may be possible to significantly reduce the complexity of deployment via the help of some built-in conan facilities. 
+Once you have your project built with Conan packages you can extract the needed artifacts to a local directory. Check out 
```
```suggestion
If you have a project with all the dependencies managed with Conan and you want to deploy it the process is:

- Extract the needed artifacts to a local directory. Check out the section :ref:`using_deploy_generator` and :ref:`using_json_generator` in order to get information on how to extract build artifacts.  
- Convert the artifacts (typically executables, shared libraries and assets) to a different deploy format. In the following sections, we are going to review some of the deploy technologies.

```

SSE4 commented:
subsections added

danimtb commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,214 @@ with:
+Using deploy generator
+----------------------
+
+With the help of the :ref:`deploy generator <deploy_generator>`, it's possible to have all dependencies of the application to be copied into 
```
Too complex sentence. Please rephrase: `The deploy generator is used to have all the dependencies of an application copied into a single place. Then all the files can be repackaged into the distribution format of choice.`

SSE4 replied:
done

danimtb commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,214 @@ with:
+
+    $ conan install . -g deploy
+
+This helps to collect all the dependencies into a single place, moving them out of the conan cache directory.
```
```suggestion
This helps to collect all the dependencies into a single place, moving them out of the Conan cache.
```

danimtb commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,214 @@ with:
 
+.. _using_deploy_generator:
+
+Using deploy generator
```
```suggestion
Using the `deploy` generator
```

danimtb commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,214 @@ with:
+
+.. _using_json_generator:
+
+Using json generator
```
```suggestion
Using the `json` generator
```

danimtb commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,214 @@ with:
+
+This helps to collect all the dependencies into a single place, moving them out of the conan cache directory.
+
+.. _using_json_generator:
```
This reference is very general
```suggestion
.. _deployable_json_generator:
```

danimtb commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,214 @@ with:
 
     Read more about the :ref:`deploy() <method_deploy>` method.
 
+.. _using_deploy_generator:
```
```suggestion
.. _deployable_deploy_generator:
```

danimtb commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,214 @@ with:
+--------------------
+
+A more advanced approach is to use the :ref:`json generator <json_generator>`:
+file containing all required information about dependencies:
```
```suggestion
This generator works in a similar fashion as the `deploy` one, although it doesn't copy the files to a directory. Instead, it generates a JSON file with all the information about the dependencies including the location of the files in the Conan cache.
```

danimtb commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,214 @@ with:
+
+    $ conan install . -g json
+
+The `conanbuildinfo.json` file produced is fully machine-readable and could be used by scripts to prepare the distribution.
```
```suggestion
The *conanbuildinfo.json* file produced is fully machine-readable and could be used by scripts to prepare the files and recreate the appropriate format for distribution.
```

danimtb commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,214 @@ with:
+    $ conan install . -g json
+
+The `conanbuildinfo.json` file produced is fully machine-readable and could be used by scripts to prepare the distribution.
+
```
Please give context about the following code snippet

danimtb commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,214 @@ with:
+                    bin_dir = os.path.relpath(bin_path, root)
+                    dep_bin_dirs[bin_path] = bin_dir
+
+The advantage over the ``deploy`` generator is fine-grained control: here we copy only binaries and libraries.
```
```suggestion
While with the `deploy` generator all the files were copied into a folder, the advantage with the `json` one is that you have fine-grained control over the files and those can be directly copied to the desired layout.
```

danimtb commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,214 @@ with:
+                    dep_bin_dirs[bin_path] = bin_dir
+
+The advantage over the ``deploy`` generator is fine-grained control: here we copy only binaries and libraries.
+It's also could be easily modified to apply some sort of filtering (e.g. to copy only shared libraries, 
```
```suggestion
In that sense, the script above could be easily modified to apply some sort of filtering (e.g. to copy only shared libraries, 
```

danimtb commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,214 @@ with:
+The advantage over the ``deploy`` generator is fine-grained control: here we copy only binaries and libraries.
+It's also could be easily modified to apply some sort of filtering (e.g. to copy only shared libraries, 
+and omit any static libraries or auxiliary files such as pkg-config .pc files).
+The extracted information may also be used to generate a simple startup script (as described below):
```
```suggestion

Additionally, you could also write a simple startup script for your application with the extracted information like this:
```

danimtb commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,214 @@ with:
+
+.. note::
+
+    The full example might be found on `GitHub <https://github.com/SSE4/conan-deploy-tool>`_.
```
examples should go to the conan-io/examples repository

SSE4 replied:
opened https://github.com/conan-io/examples/pull/36

danimtb commented on devtools/running_packages.rst:
```diff
@@ -114,6 +113,214 @@ with:
+
+    The full example might be found on `GitHub <https://github.com/SSE4/conan-deploy-tool>`_.
+
+.. _deployment_challenges:
```
IMO this part does not fit very well in here. But I am not sure where we should put this

SSE4 replied:
I am open for suggestions

danimtb commented:
It seems I missed the `deployment_challenges` label. Please fix it when possible.

Also, remember to address the points described in sse4/docs#1

SSE4 commented:
@danimtb 
- I've fixed the label
- added more background information about appimage/flatpak/snap (please lemme know if it's enough or you need to more background and what kind? I personally don't want to have text too large to read)
- I've decided to remove others section completely (puppet/chef/ansible)
- added clarifications about libc, libstdc++, libgcc
I am going to continue adding some missing pieces (e.g. instructions), but check if added things are good enough, if possible

danimtb commented on integrations/deployment/flatpak.rst:
```diff
@@ -19,4 +19,47 @@ The `packaging process <http://docs.flatpak.org/en/latest/first-build.html>`__ i
 - Run the ``flatpak-builder`` in order to produce the application
 - `publish <http://docs.flatpak.org/en/latest/publishing.html>`__ the application for further distribution
 
+With help of conan's :ref:`json generator<deployable_json_generator>`, the manifest creation could be easily automated. For example, the custom script could generate ``build-commands`` and ``sources`` entries within the manifest file:
```
Please add a link to the manifest file format or to a reference page in the flatpak's docs

SSE4 replied:
added a link to added a link to http://docs.flatpak.org/en/latest/manifests.html

danimtb commented on integrations/deployment.rst:
```diff
@@ -0,0 +1,123 @@
+    /hello: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.26' not found (required by /hello)
+
+Fortunately, this is much easier to address by just adding ``-static-libstdc++`` compiler flag. Unlike C runtime, C++ runtime can be 
+linked statically safely, because it doesn't use system calls directly, but instead relies on ``libc`` to provide required wrappers.
```
👍 much better

danimtb commented on integrations/deployment/flatpak.rst:
```diff
@@ -0,0 +1,65 @@
+- Install the flatpak runtime, flatpak-builder and SDK.
+- Create a manifest ``<app-id>.json``
+- Run the ``flatpak-builder`` in order to produce the application
+- `publish <http://docs.flatpak.org/en/latest/publishing.html>`__ the application for further distribution
```
```suggestion
- `Publish <http://docs.flatpak.org/en/latest/publishing.html>`__ the application for further distribution
```

danimtb commented on integrations/deployment/makeself.rst:
```diff
@@ -0,0 +1,23 @@
+Check out the `guide <http://xmodulo.com/how-to-create-a-self-extracting-archive-or-installer-in-linux.html>`_ on how to make
+self-extracting archive with Makeself.
+
+With help of :ref:`deploy generator <deployable_deploy_generator>`, it's only needed to invoke ``makeself.sh`` in order to generate 
```
As this seems like an easy process: Is it possible to add the commands one should execute to get the standalone app?

SSE4 replied:
sure, I am adding it!

SSE4 replied:
added a link to http://docs.flatpak.org/en/latest/manifests.html

SSE4 replied:
added actual commands for makeself

danimtb commented:
Looking better now and really like the explanations on the challenges section. Still, think the integrations could be more practical with code snippets and steps to execute, but it is a good starting point
